cmd/promtool/backfill_test.go

```diff
@@ -0,0 +1,343 @@
+		sx, sy := samples[x], samples[y]
+		if sx.Timestamp != sy.Timestamp {
+			return sx.Timestamp < sy.Timestamp
+		}
+		return sx.Value < sy.Value
+	})
+}
+
+func queryblock(t testing.TB, q storage.Querier, lbls labels.Labels, matchers ...*labels.Matcher) ([]backfillSample, error) {
+	ss := q.Select(false, nil, matchers...)
+	defer func() {
+		require.NoError(t, q.Close())
+	}()
+	samples := []backfillSample{}
+	for ss.Next() {
+		series := ss.At()
+		it := series.Iterator()
+		for it.Next() {
+			ts, v := it.At()
+			samples = append(samples, backfillSample{Timestamp: ts, Value: v, Labels: series.Labels()})
+		}
+		require.NoError(t, it.Err())
+		if len(samples) == 0 {
+			continue
+		}
+	}
+	return samples, nil
+}
+
+func testBlocks(t *testing.T, blocks []tsdb.BlockReader, expectedMinTime, expectedMaxTime int64, expectedSamples []backfillSample, metricLabels []string, expectedSymbols []string, expectedNumBlocks int) {
+	require.Equal(t, expectedNumBlocks, len(blocks))
+	allSymbols := make(map[string]struct{})
+	allSamples := make([]backfillSample, 0)
+	var maxt, mint int64 = math.MinInt64, math.MaxInt64
+	for _, block := range blocks {
+		err := func() error {
+			index, err := block.Index()
+			defer func() {
+				err = index.Close()
+				require.NoError(t, err)
+			}()
+			if maxt < block.Meta().MaxTime {
+				maxt = block.Meta().MaxTime
+			}
+			if mint > block.Meta().MinTime {
+				mint = block.Meta().MinTime
+			}
+			require.NoError(t, err)
+			symbols := index.Symbols()
+			for symbols.Next() {
+				key := symbols.At()
+				if _, ok := allSymbols[key]; !ok {
+					allSymbols[key] = struct{}{}
+				}
+			}
+			q, err := tsdb.NewBlockQuerier(block, math.MinInt64, math.MaxInt64)
+			require.NoError(t, err)
+			series, err := queryblock(t, q, labels.FromStrings(metricLabels...), labels.MustNewMatcher(labels.MatchRegexp, "", ".*"))
+			require.NoError(t, err)
+			allSamples = append(allSamples, series[0])
+			return nil
+		}()
+		require.NoError(t, err)
+	}
+	allSymbolsSlice := make([]string, 0)
+	for key := range allSymbols {
+		allSymbolsSlice = append(allSymbolsSlice, key)
+	}
+	sort.Strings(allSymbolsSlice)
+	sort.Strings(expectedSymbols)
+	require.Equal(t, expectedSymbols, allSymbolsSlice)
+	sortSamples(allSamples)
+	sortSamples(expectedSamples)
+	require.Equal(t, expectedSamples, allSamples)
+	require.Equal(t, expectedMinTime, mint)
+	require.Equal(t, expectedMaxTime, maxt)
+}
+
+func TestBackfill(t *testing.T) {
+	tests := []struct {
+		ToParse      string
+		IsOk         bool
+		MetricLabels []string
+		Description  string
+		Expected     struct {
+			MinTime   int64
+			MaxTime   int64
+			Symbols   []string
+			NumBlocks int
+			Samples   []backfillSample
+		}
+	}{
+		{
+			ToParse:     `# EOF`,
+			IsOk:        true,
+			Description: "Empty file.",
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1565133713989
```

Comment:
These timestamps look to be in milliseconds, which could cause confusion for anyone looking through this code. We should stick with realistic values in seconds.